### PR TITLE
Replace scoped_enum with a better version

### LINKED ISSUE #186: `@scoped_enum` should handle words ending with `y`
I'm opening the issue to track a feature request.

### PR DESCRIPTION
**Pull request by daniel-thom**
Fixes #186 

Please don't merge until we agree on companion PRs in the other repos.

### CODE REVIEW

File: src/utils/utils.jl [hunk 1]
```diff
@@ -192,30 +192,56 @@ function compare_values(x::T, y::U)::Bool where {T, U}
...
+            JSON3.StructType(::Type{$T}) = JSON3.StructTypes.StringType()
+
+            function _itr(res)
+                isnothing(res) && return res
+                value, state = res
+                return ($T(value), state)
+            end
+            Base.iterate(::Type{$T}) = _itr(iterate(keys(VALUE2NAME)))
+            Base.iterate(::Type{$T}, state) = _itr(iterate(keys(VALUE2NAME), state))
+            Base.length(::Type{$T}) = mapreduce(x -> 1, +, (x for x in $T))
```
Comment (daniel-thom) [inline src/utils/utils.jl:239]: @kdheepak Would you like to see any changes here? I did add this line to make it easier to collect all instances.

Comment (kdheepak) [reply src/utils/utils.jl:239]: If you implement `Base.isless(x::T, y::T)` it'll let a user be able to compare and deal with ordering. Do we think that could be useful?

Comment (kdheepak) [reply src/utils/utils.jl:244]: Thoughts on generating this in the macro?

```
Base.convert(::Type{Fruit}, x::Integer) = Fruit(x)
```

It'll allow for functionality like this:

Screen Shot 2021-01-15 at 5 08 33 PM

Comment (kdheepak) [reply src/utils/utils.jl:244]: I'm not sure about using iterate on the `type`. I'd prefer implementing some function (like `pairs`) that the user can then iterate over or call length on.

Comment (daniel-thom) [reply src/utils/utils.jl:244]: This already works
```
julia> Fruit(1)
Fruit.APPLE = 1
```
The struct version does not already work. I'm not opposed to adding it. My initial reaction is that it defeats the point of having strongly-typed enums.

I don't think we need/want ordering. Hopefully, users don't care about comparisons of values. If a user does want ordering, do they want it ordered by name (alphabetical) or by value?

They can always do one of these:
```
sort!([Fruit.APPLE, Fruit.ORANGE], by = x -> x.value)
sort!([Fruit.APPLE, Fruit.ORANGE], by = x -> string(x))
```

Comment (daniel-thom) [reply src/utils/utils.jl:244]: Responding to `iterate`. The original author had that. It allows this:
```
julia> for x in Fruit
       @show x
       end
x = Fruit.ORANGE = 2
x = Fruit.APPLE = 1
```
It did not allow for `[x for x in Fruit]`. Adding `Base.length` makes that work.

`pairs` implies that the user wants the value, right? One of my underlying assumptions is that the value should not matter. Things could be problematic if that does matter and we remove an entry.

Comment (kdheepak) [reply src/utils/utils.jl:244]: Summary, we decided to implement `isless` and not implement `length` or `iterate`. User can use `instances` to get all instances of an Scoped Enum, similar to how they would in `Base`.

Comment (daniel-thom) [reply src/utils/utils.jl]: Implemented `isless` and `instances`, deleted `iterate` and `length`, as per our discussion.

**Review by jd-lara — APPROVED** [review]
Awesome improvement

File: test/test_utils.jl [hunk 2]
```diff
@@ -28,14 +28,17 @@ end
...
-    ORANGE
-end
+IS.@scoped_enum Fruit APPLE = 1 ORANGE = 2
 
 @testset "Test scoped_enum" begin
-    @test Fruits.APPLE isa Fruits.Fruit
-    @test Fruits.ORANGE isa Fruits.Fruit
+    @test Fruit.APPLE isa Fruit
+    @test Fruit.ORANGE isa Fruit
+    sort([Fruit.ORANGE, Fruit.APPLE]) == [Fruit.APPLE, Fruit.ORANGE]
```
Comment (raphaelsaavedra) [inline test/test_utils.jl:36]: Is there a `@test` missing?

```suggestion
    @test sort([Fruit.ORANGE, Fruit.APPLE]) == [Fruit.APPLE, Fruit.ORANGE]
```

Comment (daniel-thom) [reply test/test_utils.jl]: Oops. Thanks!

**Pull request merged**